### PR TITLE
JP-2681 add residual_fringe to calwebb_spec2 pipeline 

### PR DESCRIPTION
**Pull request by jemorrison**
Resolves [JP-2681](https://jira.stsci.edu/browse/JP-2681)

This PR incorporates the MRS residual fringe correction into the calwebb_spec2 pipeline. 

**Checklist for maintainers**
- [x] added entry in `CHANGES.rst` within the relevant release section
- [x] updated or added relevant tests
- [x] updated relevant documentation
- [x] added relevant milestone
- [x] added relevant label(s)
- [x] ran regression tests, post a link to the Jenkins job below.
      [How to run regression tests on a PR](https://github.com/spacetelescope/jwst/wiki/Running-Regression-Tests-Against-PR-Branches)
https://plwishmaster.stsci.edu:8081/job/RT/job/JWST-Developers-Pull-Requests/439/testReport/junit/jwst.regtest/test_miri_residual_fringe/_stable_deps__test_residual_fringe_cal/

- [ ] Make sure the JIRA ticket is [resolved properly](https://github.com/spacetelescope/jwst/wiki/How-to-resolve-JIRA-issues)

### CODE REVIEW

**Comment by drlaw1558** [comment]
Will try to take a look at this tomorrow; hopefully updating the calspec2 param ref file won't crash the old pipeline adding a step it doesn't expect.  Might be good to update the regression test data to flight at the same time- would suggest using some 2C data from program 1524.

**Comment by jemorrison** [comment]
Ok thanks I will grab some of the new flight data and update the regression tests.

**Comment by hbushouse** [comment]
> Will try to take a look at this tomorrow; hopefully updating the calspec2 param ref file won't crash the old pipeline adding a step it doesn't expect. Might be good to update the regression test data to flight at the same time- would suggest using some 2C data from program 1524.

Oh, that could be a problem (updating the pars ref file with a new step name). I just did a quick experiment of running "strun calwebb_spec2 jw_blee_rate.fits --steps.residual_fringe.skip=True" and it threw an error for "unrecognized arguments". 

You can still go ahead and deliver a new pars ref file, but we'll have to coordinate the activation of the CRDS context that it's in with the release of the pipeline version that contains the step. So the new pars ref files definitely can NOT be delivered to Ops any time soon and should also not even go to CRDS-PUB, because theoretically PUB should support only stuff that's actually publicly released. So this would have to go to CRDS-TEST and then deployed to PUB when we release Cal for B9.0 (I think?). Should check with ReDCaT folks about the proper logistics for this.

**Comment by drlaw1558** [comment]
Yep, new param ref file crashes the current pipeline in my tests as well.  Rather than waiting until B9.0 and worrying about coordinating updates to avoid crashes it sounds much cleaner and easier to do as you suggest and hardwire skip=True, at least for the time being.  That way we can make the code available for interested users much sooner as well.  I can simply deliver a new param ref file after the next build comes out.

**Comment by jemorrison** [comment]
in the spec of the step I set:
 skip = boolean(default = True)

But when I run it on a file the residual fringe step still runs. I there something else I need to do ?

**Comment by hbushouse** [comment]
That should do it, just like we do for `wfss_contam` at https://github.com/spacetelescope/jwst/blob/master/jwst/wfss_contam/wfss_contam_step.py#L22

Is there anything else, like in a pars ref file or something, that might be overriding it?

**Comment by drlaw1558** [comment]
One oddity that I've noticed on testing this branch; if residual_fringe.skip is True, there's no logging information written out about how the step was skipped, and no FITS header keywords saying that it was skipped.  This is different from the performance I'm used to from other steps, which usually start the step, prints a log message about skipping the step, and sets SKIPPED into the relevant S_ header keyword.

If residual_fringe.skip is False then the step runs as expected, prints logging information and populates header keywords.  There is, however, an unusual log message in a different font/color from all other log messages (white background in a Notebook run instead of the usual pink):
Type of calibrated <class 'jwst.datamodels.ifuimage.IFUImageModel'>
Why is this message done via 'print' instead of self.log ?

**Comment by jemorrison** [comment]
David the strange print message was something I forgot to remove. It is not there any more. 
I am trying to figure out why when the step is skipped it does not write any log information or write the information to the header that the step is skipped.

**Comment by jemorrison** [comment]
@hbushouse Take a look at how I set up the residual_fringe in the table for calwebb_spec2.rst. I wanted to make clear that is was skipped so I added a subscript 2 and then a note for this subscript that the step by default is skipped.

File: CHANGES.rst
Comment (hbushouse) [inline CHANGES.rst:34]: Need to reference the PR number at the end of the entry.

File: docs/jwst/pipeline/calwebb_spec2.rst
Comment (hbushouse) [inline docs/jwst/pipeline/calwebb_spec2.rst:70]: I thought I remembered hearing that `residual_fringe` had to be done after the `photom` step, because it was setup to work on flux-calibrated data?

File: docs/jwst/residual_fringe/main.rst
Comment (hbushouse) [inline docs/jwst/residual_fringe/main.rst:27]: Is it still true that this step can accept an ASN table with a list of input files?  Or just a single input file/datamodel?

File: jwst/pipeline/calwebb_spec2.py
Comment (hbushouse) [inline jwst/pipeline/calwebb_spec2.py:283]: Looks to me like the step is being called up in line 257, so why is the comment down here?

File: jwst/residual_fringe/residual_fringe_step.py
Comment (hbushouse) [inline jwst/residual_fringe/residual_fringe_step.py:106]: Minor nit, but technically neither of the reference files is being "opened" by these calls. All you're doing is finding the names of the right files to use. They get opened later on.

**Review by hbushouse — COMMENTED** [review]
Just a few minor comments and questions.

File: docs/jwst/pipeline/calwebb_spec2.rst
Comment (jemorrison) [reply docs/jwst/pipeline/calwebb_spec2.rst:70]: you are right. I fixed the table. It is in the correct place when it is called in the pipeline

File: jwst/pipeline/calwebb_spec2.py
Comment (jemorrison) [reply jwst/pipeline/calwebb_spec2.py:283]: Forgot to move comment when I moved where the code was run

**Comment by jemorrison** [comment]
@drlaw1558 Try it now. You should now have the S_RESFRI=SKIPPED by default and you should also have logging information printed to the screen saying it is skipping the step.

**Comment by drlaw1558** [comment]
I can see the expected logging information on the step now, both when run and skipped.  Still seems to be missed information about the reference file used though; I get the expected CRDS prefetch information but the FITS header doesn't seem to set the R_RESFRI or R_FRIFRQ (whatever the keyword is meant to be) giving the name of the FRINGEFREQ reference file used in the run.

Likewise, the logging information should probably say 'Using FRINGEFREQ reference file:' instead of 'Using Residual FRINGE reference file:'

File: jwst/pipeline/calwebb_spec2.py
Comment (hbushouse) [inline jwst/pipeline/calwebb_spec2.py:257]: I wonder if it would be a little cleaner (in terms of code complexity) to move this call into the `_process_common` function, along with the `fringe` step. The setting of the `skip` param will still control whether or not it gets applied to data for different modes.

Comment (jemorrison) [reply jwst/pipeline/calwebb_spec2.py:257]: David - currently there is no keyword for the FRINGEFREQ in the core schema. I need to make up one. Do you have a preference R_FRIFRQ ?

Comment (drlaw1558) [reply jwst/pipeline/calwebb_spec2.py:257]: R_FRIFRQ makes sense to me, but I defer to whatever you and Howard think most consistent with the other keywords.

Comment (jemorrison) [reply jwst/pipeline/calwebb_spec2.py:257]: @drlaw1558 have you used the parameters 'ignore_region_min' or 'ignore_region_max' to not used wavelengths. I can not get these parameters to work anymore

Comment (drlaw1558) [reply jwst/pipeline/calwebb_spec2.py]: Can't say as I've ever tried using those before

**Review by drlaw1558 — APPROVED** [review]
Looks good to me!  Seeing the expected performance and logging information.

**Comment by jemorrison** [comment]
one of the unit tests fails now because there is a error in stpipe how it works with lists as input parameters.https://github.com/spacetelescope/stpipe/issues/71

Hopefully we can still merge with this error - or I could comment out the test until the fix is made.

**Comment by jemorrison** [comment]
@hbushouse 
I put the link to the regression test. The difference in the residual_fringe tests is now S_RESFRI and R_FRIFRQ. Is written. I don't think the other differences matter.  Should I update the truth file ? I am not sure how I get access to the one created with this regression test. Or should we just wait and Okify them ?
I think all the other CI failures are related to the error about not being about the parse the list parameter - which Mihai already filed a ticket for.

**Comment by hbushouse** [comment]
Once the latest run of the CI tests has finished (hoping that a few more of them might succeed this time around) I think it'll be ready to merge. I agree that the regtest results are OK. We'll okify all the truth files after we've merged and done another regular full regtest run.

**Comment by hbushouse** [comment]
At least one of the CI tests is showing an error in jwst/jwst/residual_fringe/tests/test_configuration.py

**Comment by jemorrison** [comment]
Yes this test is where I discovered that there is now a bug in parsing a parameter with a list. I can comment out that test until that gets fixed if you want

**Comment by hbushouse** [comment]
> Yes this test is where I discovered that there is now a bug in parsing a parameter with a list. I can comment out that test until that gets fixed if you want

Just use the pytest.mark.xfail command to mark it as failing for now.

**Comment by jemorrison** [comment]
Added pytest.mark.xfail to test. This PR should be ready to go

**Pull request merged**
